### PR TITLE
[framework] governance multi-step proposal

### PR DESCRIPTION
**Pull request by 0xchloe**
### Description
Implementation of this AIP: https://github.com/aptos-foundation/AIPs/issues/3. 

TLDR: we're supporting multi-step proposals in this PR. We do so by updating `proposal.execution_hash` and `ApprovedExecutionHashes` on-chain to the `next_execution_hash`. 

Summary of changes included in this PR:

`aptos_governance` module: 
Added:
- `aptos_governance::create_proposal_v2()`: this is similar to the existing `aptos_governance::create_proposal`, but here we call the newly added `voting::create_proposal_v2` to create an Aptos governance proposal. 

- `aptos_governance::resolve_multi_step_proposal()`: similar to the existing `aptos_governance::resolve()`, but here we replace the current execution hash in `ApprovedExecutionHashes` with `next_execution_hash`, and call `voting::resolve_proposal_v2` to resolve the proposal. 

Modified:
- `aptos_governance::add_approved_script_hash()`: previously, we immediately return if the `ApprovedExecutionHashes` map already contains the given `proposal_id`. In the updated code, if the `ApprovedExecutionHashes` map already contains the given `proposal_id`, we will replace the existing value in `ApprovedExecutionHashes` with the updated execution hash.  

- `aptos_governance::create_proposal()`: redirected this function to call the newly added `aptos_governance::create_proposal_v2()`, so that we'll have metadata information about if this proposal is single-step or multi-step. 

`voting` module: 
Added:
- Two new constants: `IS_MULTI_STEP_PROPOSAL_KEY` and `IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY`. They will be stored in the `proposal.metadata` map upon creation of a proposal. 

- `voting:: create_proposal_v2()`: the flow of this function is similar to the existing `voting:: create_proposal()` function except that we are also adding metadata about this proposal in the `proposal.metadata` map. 
     - `IS_MULTI_STEP_PROPOSAL_KEY`: `true` for multi-step proposals, `false` for single-step proposals;
     - `IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY` only exists in multi-step proposals. Its value is by default `false`. 
         We update it to true when we start resolving the multi-step proposal. This flag is used to disable further voting when a multi-step proposal is already in execution. 

- `voting:: is_proposal_resolvable()`: this includes all common checks on if a proposal is resolvable regardless if the proposal is single-step or multi-step. 

- `voting:: resolve_proposal_v2()`: this function can resolve a single-step proposal or a multi-step proposal. The flow for a single-step proposal remains mostly unchanged. The flow for a multi-step proposal will update the `proposal.execution_hash` with the `next_execution_hash` on-chain, and check/update `proposal.metadata` values. 

- `voting:: is_multi_step_proposal_in_execution()`: this function returns true if the specified multi-step proposal is in execution. 

Modified:
- `voting::vote()`: added a check to abort when the specified proposal is multi-step and already in execution. 

- `voting::resolve()`: added a check to ensure that this function can only be called to resolve single-step proposals.

### Test Plan

Unit tests in this PR, smoke + e2e tests to come in future PRs. 

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/5445)

### CODE REVIEW

**Review by davidiw — COMMENTED** [review]
ahh this is clever!

File: aptos-move/framework/aptos-framework/sources/voting.move [hunk 1]
```diff
@@ -299,6 +309,9 @@ module aptos_framework::voting {
         let proposal = table::borrow_mut(&mut voting_forum.proposals, proposal_id);
         assert!(!proposal.is_resolved, error::invalid_state(EPROPOSAL_ALREADY_RESOLVED));
 
+        let multi_step_key = utf8(IS_MULTI_STEP_PROPOSAL_KEY);
+        assert!(!simple_map::contains_key(&proposal.metadata, &multi_step_key) || *simple_map::borrow(&proposal.metadata, &multi_step_key) == SINGLE_STEP_PROPOSAL, error::unauthenticated(EPROPOSAL_IS_NOT_SINGLE_STEP));
```
Comment (runtian-zhou) [inline aptos-move/framework/aptos-framework/sources/voting.move:313]: Any particular reason why we have one table but with two types of keys instead of two tables? Does that has anything to do with backwards compatibility?

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 2]
```diff
@@ -274,6 +285,81 @@ module aptos_framework::aptos_governance {
         );
     }
 
+    /// Create a proposal with the backing `stake_pool`.
+  /// @param execution_hash Required. This is the hash of the resolution script. When the proposal is resolved,
+  /// only the exact script with matching hash can be successfully executed.
+    public entry fun create_proposal_generic(
```
Comment (runtian-zhou) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:291]: For future compatibility consideration maybe `create_multi_proposal` is better? My concern is in the future if we want to add new functionailities here we won't be able to use `create_proposal_generic` any more because of the upgrade checks.

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 1 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move:313]: Yeah it is for backward compatibility; existing (single-step) proposals wouldn't have the `multi_step_key` in their `proposal.metadata`. So we would assume a proposal is single-step if their simple map does not have the `multi_step_key`, or if the value is `SINGLE_STEP_PROPOSAL` to their `multi_step_key`.

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move (hunk 2 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/aptos_governance.move:291]: oooh I see, will update the naming : )

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 3]
```diff
@@ -238,6 +263,12 @@ module aptos_framework::aptos_governance {
         // Create and validate proposal metadata.
         let proposal_metadata = create_proposal_metadata(metadata_location, metadata_hash);
 
+        if (is_multi_step_proposal) {
+            add_is_multi_step_proposal_flag(&mut proposal_metadata, MULTI_STEP_PROPOSAL);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:267]: Why not just inline the simple_map add?

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 4]
```diff
@@ -421,27 +486,103 @@ module aptos_framework::aptos_governance {
         metadata
     }
 
+    fun add_is_multi_step_proposal_flag(metadata: &mut SimpleMap<String, vector<u8>>, is_multi_step_proposal_value: vector<u8>) {
+        assert!(is_multi_step_proposal_value == MULTI_STEP_PROPOSAL || is_multi_step_proposal_value == SINGLE_STEP_PROPOSAL, error::invalid_argument(EPROPOSAL_METADATA_INVALID_KEY));
+        simple_map::add(metadata, utf8(IS_MULTI_STEP_PROPOSAL_KEY), is_multi_step_proposal_value);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:491]: Why not just a simple boolean instead of vec<u8> for is_multi_step_proposal_value?

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 5]
```diff
@@ -354,6 +385,33 @@ module aptos_framework::aptos_governance {
         simple_map::add(&mut approved_hashes.hashes, proposal_id, execution_hash);
     }
 
+    /// Used when making a multi-step governance proposal.
+    /// Replace the current execution hash in the `ApprovedExecutionHashes` with the next execution hash
+    fun replace_approved_hash(proposal_id: u64, next_execution_hash: vector<u8>) acquires ApprovedExecutionHashes {
+        let approved_hashes = borrow_global_mut<ApprovedExecutionHashes>(@aptos_framework);
+        if (!simple_map::contains_key(&approved_hashes.hashes, &proposal_id)) {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:392]: When does this happen?

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 6]
```diff
@@ -354,6 +385,33 @@ module aptos_framework::aptos_governance {
...
+            return
+        };
+
+        let proposal_state = voting::get_proposal_state<GovernanceProposal>(@aptos_framework, proposal_id);
+        assert!(proposal_state == PROPOSAL_STATE_SUCCEEDED, error::invalid_argument(EPROPOSAL_NOT_RESOLVABLE_YET));
+
+        simple_map::remove(&mut approved_hashes.hashes, &proposal_id);
+        if (vector::length(&next_execution_hash)!=0) {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:400]: nit: spaces around !=

File: aptos-move/framework/aptos-framework/sources/voting.move [hunk 7]
```diff
@@ -326,6 +339,55 @@ module aptos_framework::voting {
...
+    /// Resolve the multi-step proposal with given id. Can only be done if there are at least as many votes as min required and
+    /// there are more yes votes than no. If either of these conditions is not met, this will revert.
+    ///
+    /// @param voting_forum_address The address of the forum where the proposals are stored.
+    /// @param proposal_id The proposal id.
+    /// @param next_execution_hash The next execution hash of this multi-step proposal
+    public fun resolve_multi_step_proposal<ProposalType: store>(
+        voting_forum_address: address,
+        proposal_id: u64,
+        curr_execution_hash: vector<u8>,
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/voting.move:351]: Why is this passed as an argument? Shouldn't this be stored in the proposal? If you can pass any curr_execution_hash, wouldn't you be able to execute anything?

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 8]
```diff
@@ -238,6 +263,12 @@ module aptos_framework::aptos_governance {
         // Create and validate proposal metadata.
         let proposal_metadata = create_proposal_metadata(metadata_location, metadata_hash);
 
+        if (is_multi_step_proposal) {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:266]: Why not set this multistep flag inside voting.move instead of here? This could be cleaner since you wouldn't need to duplicate the constants in both modules.

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 1 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/voting.move:313]: Why not just use a boolean and interpret the lack of this flag as single-step? This should maintain backward compatibility

Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move:313]: ah so it's bc we  made the metadate simple map to be of type `SimpleMap<String, vector<u8>>` [here](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/aptos_governance.move#L414)

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move (hunk 3 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/aptos_governance.move:267]: good point, updating now : )

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move (hunk 4 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/aptos_governance.move:491]: it's bc the metadata map is of type `<String, vector<u8>>`

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move (hunk 6 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/aptos_governance.move:400]: good catch, updating now : )

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move (hunk 8 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/aptos_governance.move:266]: I was thinking that if we wanted to set the flag inside `voting.move`, we'd need to add a new create-proposal function to take in an additional `is_multi_step_proposal`. By adding it in the proposal metadata map within the `aptos_governance` module, we implicitly pass in the `is_multi_step_proposal` information without having to add a new function. But we can certainly add a new function in `voting.move` if that's cleaner

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 1 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move:313]: btw we're interpreting the lack of this flag as single-step in the code above

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move (hunk 5 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/aptos_governance.move:392]: good point, deleting the check

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 7 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move:351]: good point, updating now : )

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move (hunk 8 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/aptos_governance.move:266]: I think adding a new function in voting module would be cleaner.

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 1 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/voting.move:313]: We can just convert back and forth between vector<u8> and bool right?

Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move:313]: oh yeah makes sense, I will use a boolean and convert it to vector<u8> then

File: aptos-move/framework/aptos-framework/sources/voting.move [hunk 9]
```diff
@@ -193,6 +202,30 @@ module aptos_framework::voting {
...
+            execution_hash,
+            min_vote_threshold,
+            expiration_secs,
+            early_resolution_vote_threshold,
+            metadata,
+            false
+        )
+    }
+
+    public fun create_single_step_or_multi_step_proposal<ProposalType: store>(
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/voting.move:219]: Can you add some docs for this function, similar to create_proposal?

File: aptos-move/framework/aptos-framework/sources/voting.move [hunk 10]
```diff
@@ -299,6 +343,9 @@ module aptos_framework::voting {
         let proposal = table::borrow_mut(&mut voting_forum.proposals, proposal_id);
         assert!(!proposal.is_resolved, error::invalid_state(EPROPOSAL_ALREADY_RESOLVED));
 
+        let multi_step_key = utf8(IS_MULTI_STEP_PROPOSAL_KEY);
+        assert!(!simple_map::contains_key(&proposal.metadata, &multi_step_key) || *simple_map::borrow(&proposal.metadata, &multi_step_key) == to_bytes(&false), error::permission_denied(EMULTI_STEP_PROPOSAL_CANNOT_USE_SINGLE_STEP_RESOLVE_FUNCTION));
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/voting.move:347]: nit: This is a bit hard to read. We can declare two variables here to improve readability:

// Default to treating the proposal as single-step if the is_multi_step_proposal attribute is not present.
let has_multi_step_attribute = !simple_map::contains_key(&proposal.metadata, &multi_step_key);
let is_multi_step_proposal = from_bcs::to_bool(*simple_map::borrow(&proposal.metadata, &multi_step_key));
assert!(!has_multi_step_attribute || !is_multi_step_proposal, ...);

File: aptos-move/framework/aptos-framework/sources/voting.move [hunk 11]
```diff
@@ -326,6 +373,67 @@ module aptos_framework::voting {
...
+        let proposal_state = get_proposal_state<ProposalType>(voting_forum_address, proposal_id);
+        assert!(proposal_state == PROPOSAL_STATE_SUCCEEDED, error::invalid_state(EPROPOSAL_CANNOT_BE_RESOLVED));
+
+        let voting_forum = borrow_global_mut<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = table::borrow_mut(&mut voting_forum.proposals, proposal_id);
+        assert!(!proposal.is_resolved, error::invalid_state(EPROPOSAL_ALREADY_RESOLVED));
+
+        // If this proposal is multi-step, we update the metadata map to indicate that this proposal is in execution.
+        let multi_step_in_execution_key = utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        if (simple_map::contains_key(&proposal.metadata, &multi_step_in_execution_key)) {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/voting.move:397]: This is interesting. There's some redundancy between IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY and IS_MULTI_STEP_PROPOSAL:

1. If IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY is not present in the metadata, the proposal is single-step.
2. If IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY is present but false, the proposal is multi-step and has not entered execution phase.
3. If IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY is present but true, the proposal is multi-step and has entered execution phase.

I wonder if we can combine these 2 flags?

File: aptos-move/framework/aptos-framework/sources/voting.move [hunk 12]
```diff
@@ -207,6 +240,13 @@ module aptos_framework::voting {
         let proposal_id = voting_forum.next_proposal_id;
         voting_forum.next_proposal_id = voting_forum.next_proposal_id + 1;
 
+        // Add a flag to indicate if this proposal is single-step or multi-step.
+        simple_map::add(&mut metadata, utf8(IS_MULTI_STEP_PROPOSAL_KEY), to_bytes(&is_multi_step_proposal));
+        if (is_multi_step_proposal) {
+            // Add a flag to indicate if a multi-step proposal is in execution. If yes, we will disable voting.
+            simple_map::add(&mut metadata, utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY), to_bytes(&false));
+        };
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/voting.move:248]: Just for defensive programming, can we remove IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY if it's set in the metadata by client but is_multi_step_proposal is false? This would keep things consistent later on

File: aptos-move/framework/aptos-framework/sources/voting.move [hunk 13]
```diff
@@ -326,6 +373,67 @@ module aptos_framework::voting {
...
+        // If this proposal is multi-step, we update the metadata map to indicate that this proposal is in execution.
+        let multi_step_in_execution_key = utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        if (simple_map::contains_key(&proposal.metadata, &multi_step_in_execution_key)) {
+            let is_multi_step_proposal_in_execution_value = simple_map::borrow_mut(&mut proposal.metadata, &multi_step_in_execution_key);
+            if (!from_bcs::to_bool(*is_multi_step_proposal_in_execution_value)) {
+                *is_multi_step_proposal_in_execution_value = to_bytes(&true);
+            };
+        };
+
+        let resolvable_time = to_u64(*simple_map::borrow(&proposal.metadata, &utf8(RESOLVABLE_TIME_METADATA_KEY)));
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/voting.move:404]: nit: This seems redundant. I wonder if we can rollup a lot of these checks around whether the proposal can be resolved so we don't have to duplicate them among the single-step and multi-step proposal execution functions

File: aptos-move/framework/aptos-framework/sources/voting.move [hunk 14]
```diff
@@ -326,6 +373,67 @@ module aptos_framework::voting {
...
+
+        let voting_forum = borrow_global_mut<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = table::borrow_mut(&mut voting_forum.proposals, proposal_id);
+        assert!(!proposal.is_resolved, error::invalid_state(EPROPOSAL_ALREADY_RESOLVED));
+
+        // If this proposal is multi-step, we update the metadata map to indicate that this proposal is in execution.
+        let multi_step_in_execution_key = utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        if (simple_map::contains_key(&proposal.metadata, &multi_step_in_execution_key)) {
+            let is_multi_step_proposal_in_execution_value = simple_map::borrow_mut(&mut proposal.metadata, &multi_step_in_execution_key);
+            if (!from_bcs::to_bool(*is_multi_step_proposal_in_execution_value)) {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/voting.move:399]: Why not always set this to true? I don't think it costs more gas than the check in the if condition

Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move:399]: could you elaborate on setting which variable to true?

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 11 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move:397]: I think we can do a key `PROPOSAL_TYPE_AND_STATE` and setting it to 
- value `SINGLE_STEP` if the proposal is single-step;
- value `MULTI_STEP` if the proposal is multi-step and not in execution;
- value `MULTI_STEP_IN_EXECUTION` if the proposal is multi-step and in execution.
However this might look messier because we're introducing three new constants.

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 13 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move:404]: yeah, or we can just direct the existing single-step resolve function to call the newly added generic resolve function.  

the checks are mostly the same, it's just the event-emitting logic and updating to next execution hash logic are different. I can add a new function `check_if_proposal_is_resolvable` to do all the checks as well, but it'll be trickier, since we're borrowing from `VotingForum` to emit events, but we're also going to use `VotingForum` in the `check_if_proposal_is_resolvable()` function as well, and I don't think that's allowed so we will probably need to do something hacky.

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 11 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/voting.move:397]: That's true. I'm fine with having 2 like what we have currently as well. Can you add some documentation above the metadata field in the struct to list the special metadata attributes that are used by the voting flow? We'll have 3 in total
1. Last voting time
2. proposal type (single vs multi-step)
3. Whether the proposal is in execution state.

Speaking of which, make sure the MULTI_STEP_IN_EXECUTION field is clear too after all steps have been executed. It's not crucial to the on chain flow but can be cleaner for off chain to interpret if it's closed out properly when execution is fully complete.

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 14 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/voting.move:399]: *is_multi_step_proposal_in_execution_value = to_bytes(&true);

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 13 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/voting.move:404]: I see that's probably fine then.

File: aptos-move/framework/aptos-framework/sources/voting.move [hunk 15]
```diff
@@ -326,6 +373,67 @@ module aptos_framework::voting {
...
+        let multi_step_key = utf8(IS_MULTI_STEP_PROPOSAL_KEY);
+        let is_multi_step = simple_map::contains_key(&proposal.metadata, &multi_step_key) && *simple_map::borrow(&proposal.metadata, &multi_step_key) == to_bytes(&true);
+        // If the proposal is a single-step proposal, or if this step is the last step in a multi-step proposal,
+        // we can mark that this proposal is resolved and emit a `ResolveProposal` event.
+        if (!is_multi_step || vector::length(&next_execution_hash) == 0) {
+            let resolved_early = can_be_resolved_early(proposal);
+            proposal.is_resolved = true;
+            proposal.resolution_time_secs = timestamp::now_seconds();
+
+            event::emit_event<ResolveProposal>(
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/voting.move:421]: As discussed in the past, what do you think about emitting this event for every step in the multi-step execution? Without introducing a new event, we're not really tracking executing each step.

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 16]
```diff
@@ -362,6 +378,15 @@ module aptos_framework::aptos_governance {
         get_signer(signer_address)
     }
 
+    /// Resolve a successful multi-step proposal. This would fail if the proposal is not successful.
+    public fun resolve_multi_step_proposal(proposal_id: u64, signer_address: address, next_execution_hash: vector<u8>): signer acquires GovernanceResponsbility, ApprovedExecutionHashes {
+        voting::resolve_single_step_or_multi_step_proposal<GovernanceProposal>(@aptos_framework, proposal_id, next_execution_hash);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:383]: Don't you also need to add the hash of the next step? Otherwise, it might get rejected by mempool if exceeding the transaction size limit (64 KB)

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move [hunk 17]
```diff
@@ -215,6 +216,20 @@ module aptos_framework::aptos_governance {
...
         metadata_location: vector<u8>,
         metadata_hash: vector<u8>,
+    ) acquires GovernanceConfig, GovernanceEvents {
+        create_single_step_or_multi_step_proposal(proposer, stake_pool, execution_hash, metadata_location, metadata_hash, false);
+    }
+
+    /// Create a single-step or multi-step proposal with the backing `stake_pool`.
+    /// @param execution_hash Required. This is the hash of the resolution script. When the proposal is resolved,
+    /// only the exact script with matching hash can be successfully executed.
+    public entry fun create_single_step_or_multi_step_proposal(
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_governance.move:226]: nit: create_single_step_or_multi_step_proposal is a bit long. What do you think about create_proposal_v2? Or some other better name.

**Review by movekevin — CHANGES_REQUESTED** [review]

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 15 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move:454]: I thought we didn't want to add a new event because we'll have to do two upgrades (with the second one initializing the event struct)? Do we want to add a new `ResolveMultiStepProposal` event and where do we put that event (we cannot put that in `VotingForum`?) 

Yeah I think emitting this event for every step is simpler, just not sure if it'd be confusing to people. I was thinking that we can track each step by looking up transactions on-chain by I guess that's harder / more hassle.

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 11 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move]: Good point! Just added comments for these variables. Also cleared `MULTI_STEP_IN_EXECUTION` upon successful resolution of the multi-step proposal.

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 14 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move]: oooh I see, I thought it's cheaper to read than write (but I guess in this case, we are also converting values). just updated it to be `*is_multi_step_proposal_in_execution_value = to_bytes(&true);`

File: aptos-move/framework/aptos-framework/sources/aptos_governance.move (hunk 16 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/aptos_governance.move]: yeah good point - just updated : )

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 13 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move]: good point! I just updated this to include a `is_proposal_resolvable()` function to do all the checks because I realized that as long as we are not borrowing `VotingForum` before calling `is_proposal_resolvable()`, we will be fine.

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 14 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/voting.move]: I don't think it counts as writing if you're not changing the value?

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 15 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/voting.move:454]: Yeah ideally we would have a different event, but emitting this same event might still work? Downstream systems just need to deal with multiple resolve events for the same proposal.

File: aptos-move/framework/aptos-framework/sources/voting.move [hunk 18]
```diff
@@ -207,6 +259,23 @@ module aptos_framework::voting {
...
+        // Add a flag to indicate if this proposal is single-step or multi-step.
+        simple_map::add(&mut metadata, utf8(IS_MULTI_STEP_PROPOSAL_KEY), to_bytes(&is_multi_step_proposal));
+
+        let is_multi_step_in_execution_key = utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        if (is_multi_step_proposal) {
+            // If the proposal is a multi-step proposal, we will a flag to indicate if this multi-step proposal is in execution.
+            // If the value is true, that means that the proposal is resolving in progress, and we will disable further voting
+            // for this proposal.
+            simple_map::add(&mut metadata, is_multi_step_in_execution_key, to_bytes(&false));
+        } else {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/voting.move:271]: nit: This can be combined into else if.

File: aptos-move/framework/aptos-framework/sources/voting.move [hunk 19]
```diff
@@ -283,35 +356,59 @@ module aptos_framework::voting {
...
-        proposal.is_resolved = true;
-        proposal.resolution_time_secs = timestamp::now_seconds();
-
         assert!(
             transaction_context::get_script_hash() == proposal.execution_hash,
             error::invalid_argument(EPROPOSAL_EXECUTION_HASH_NOT_MATCHING),
         );
+    }
+
+    /// Resolve the proposal with given id. Can only be done if there are at least as many votes as min required and
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/voting.move:387]: Can you add some documentation here to explain that this can only be called for single-step proposals because multi-step proposals need the next execution hash?

File: aptos-move/framework/aptos-framework/sources/voting.move [hunk 20]
```diff
@@ -326,6 +423,54 @@ module aptos_framework::voting {
...
+                &mut voting_forum.events.resolve_proposal_events,
+                ResolveProposal {
+                    proposal_id,
+                    yes_votes: proposal.yes_votes,
+                    no_votes: proposal.no_votes,
+                    resolved_early,
+                },
+            );
+
+            if (is_multi_step) {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/voting.move:464]: This is missing the body?

**Review by movekevin — APPROVED** [review]
Just a few minor comments left. Otherwise it's good to go :)

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 15 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move:454]: ok sounds good, updating now : )

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 19 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move]: yup just added!

File: aptos-move/framework/aptos-framework/sources/voting.move (hunk 20 above)
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move:463]: good catch - just updated : ) I was using this if statement to set the `IS_MULTI_STEP_PROPOSAL_IN_EXECUTION` field back to `false` upon successful resolution of the multi-step proposal

**Review by runtian-zhou — APPROVED** [review]
Left a minor nit but the overall logic makes sense to me! Great work!

File: aptos-move/framework/aptos-framework/sources/voting.move [hunk 21]
```diff
@@ -283,35 +358,59 @@ module aptos_framework::voting {
         );
     }
 
-    /// Resolve the proposal with given id. Can only be done if there are at least as many votes as min required and
-    /// there are more yes votes than no. If either of these conditions is not met, this will revert.
-    ///
-    /// @param voting_forum_address The address of the forum where the proposals are stored.
-    /// @param proposal_id The proposal id.
-    public fun resolve<ProposalType: store>(
+    fun is_proposal_resolvable<ProposalType: store>(
```
[the remark this answers is not in the document]
Comment (0xchloe) [reply aptos-move/framework/aptos-framework/sources/voting.move:361]: great point - will update the name of this function and add more comments!
